src/docker/image.rs

```diff
@@ -0,0 +1,282 @@
+            let image_toolchain = toolchain.into();
+            let (os, arch, variant) = if let Some((os, rest)) = platform.split_once('/') {
+                let os: StrDeserializer<'_, SerdeError> = os.into_deserializer();
+                let (arch, variant) = if let Some((arch, variant)) = rest.split_once('/') {
+                    let arch: StrDeserializer<'_, SerdeError> = arch.into_deserializer();
+                    (arch, Some(variant))
+                } else {
+                    let arch: StrDeserializer<'_, SerdeError> = rest.into_deserializer();
+                    (arch, None)
+                };
+                (os, arch, variant)
+            } else {
+                eyre::bail!("invalid platform specified")
+            };
+            Ok(ImagePlatform {
+                architecture: Architecture::deserialize(arch)?,
+                os: Os::deserialize(os)?,
+                variant: variant.map(ToOwned::to_owned),
+                target: image_toolchain,
+            })
+        } else {
+            Ok(ImagePlatform::from_target(s.into())
+                .wrap_err_with(|| format!("could not map `{s}` to a platform"))?)
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Architecture {
+    I386,
+    Amd64,
+    Arm,
+    #[serde(alias = "aarch64")]
+    Arm64,
+    Mips,
+    Mips64,
+    Mips64Le,
+    MipsLe,
+    Ppc64,
+    Ppc64Le,
+    Riscv64,
+    S390x,
+    Wasm,
+}
+
+impl Architecture {
+    pub fn from_target(target: &TargetTriple) -> Result<Self> {
+        use serde::de::IntoDeserializer;
+        let arch = target
+            .triple()
+            .split_once('-')
+            .ok_or_else(|| eyre::eyre!("malformed target"))?
+            .0;
+        Self::deserialize(<&str as IntoDeserializer>::into_deserializer(arch))
+            .wrap_err_with(|| format!("architecture {arch} is not supported"))
+    }
+}
+
+impl std::fmt::Display for Architecture {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.serialize(f)
+    }
+}
+
+// Supported Oses are on
+// https://rust-lang.github.io/rustup-components-history/aarch64-unknown-linux-gnu.html
+// where rust, rustc and cargo is available (e.g rustup toolchain add works)
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Os {
+    #[serde(alias = "macos")]
+    Darwin,
+    Freebsd,
+    Illumos,
+    Linux,
+    Netbsd,
+    Windows,
+    //Aix,
+    //Android,
+    //Dragonfly,
+    //Ios,
+    //Js,
+    //Openbsd,
+    //Plan9,
+    //Solaris,
+}
+
+impl Os {
+    pub fn from_target(target: &TargetTriple) -> Result<Self> {
+        use serde::de::IntoDeserializer;
+        let vendor = target
+            .triple()
+            .split('-')
+            .nth(2)
```

Comment:
Last comment I think: doesn't this depend on if we have say, `aarch64-apple-darwin` or `x86_64-unknown-linux-gnu`? Shouldn't this be `rsplit` and then check the last 2?

Reply:
Once this is change, we're ready for a review, and I think we should merge this earlier rather than later (since it touches so much, and is well-implemented).

I've implemented the patch, which is here:

```diff
diff --git a/src/docker/image.rs b/src/docker/image.rs
index 334f9d8..3a2e064 100644
--- a/src/docker/image.rs
+++ b/src/docker/image.rs
@@ -140,6 +140,7 @@ impl std::str::FromStr for ImagePlatform {
 #[serde(rename_all = "lowercase")]
 pub enum Architecture {
     I386,
+    #[serde(alias = "x86_64")]
     Amd64,
     Arm,
     #[serde(alias = "aarch64")]
@@ -150,6 +151,7 @@ pub enum Architecture {
     MipsLe,
     Ppc64,
     Ppc64Le,
+    #[serde(alias = "riscv64gc")]
     Riscv64,
     S390x,
     Wasm,
@@ -188,26 +190,36 @@ pub enum Os {
     Netbsd,
     Windows,
     //Aix,
-    //Android,
+    Android,
     //Dragonfly,
     //Ios,
     //Js,
     //Openbsd,
     //Plan9,
-    //Solaris,
+    Solaris,
 }
 
 impl Os {
     pub fn from_target(target: &TargetTriple) -> Result<Self> {
-        use serde::de::IntoDeserializer;
-        let vendor = target
-            .triple()
-            .split('-')
-            .nth(2)
+        let mut split = target.triple().rsplit('-');
+        let rn1 = split
+            .next()
             .ok_or_else(|| eyre::eyre!("malformed target"))?;
-
-        Self::deserialize(<&str as IntoDeserializer>::into_deserializer(vendor))
-            .wrap_err_with(|| format!("os {vendor} is not supported"))
+        let rn2 = split
+            .next()
+            .ok_or_else(|| eyre::eyre!("malformed target"))?;
+        Ok(match (rn2, rn1) {
+            (_, "darwin") => Os::Darwin,
+            (_, "freebsd") => Os::Freebsd,
+            (_, "netbsd") => Os::Netbsd,
+            (_, "illumos") => Os::Illumos,
+            (_, "solaris") => Os::Solaris,
+            // android targets also set linux, so must occur first
+            (_, "android") => Os::Android,
+            ("linux", _) => Os::Linux,
+            ("windows", _) => Os::Windows,
+            _ => eyre::bail!("os {rn2} or {rn1} is not supported"),
+        })
     }
 }
 
@@ -280,3 +292,37 @@ impl ImagePlatform {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! t {
+        ($t:literal) => (TargetTriple::from($t));
+    }
+
+    #[test]
+    fn architecture_from_target() -> Result<()> {
+        assert_eq!(Architecture::from_target(&t!("x86_64-apple-darwin"))?, Architecture::Amd64);
+        assert_eq!(Architecture::from_target(&t!("arm-unknown-linux-gnueabihf"))?, Architecture::Arm);
+        assert_eq!(Architecture::from_target(&t!("aarch64-unknown-linux-gnu"))?, Architecture::Arm64);
+        assert_eq!(Architecture::from_target(&t!("mips-unknown-linux-gnu"))?, Architecture::Mips);
+        assert_eq!(Architecture::from_target(&t!("mips64-unknown-linux-gnuabi64"))?, Architecture::Mips64);
+        assert_eq!(Architecture::from_target(&t!("mips64le-unknown-linux-gnuabi64"))?, Architecture::Mips64Le);
+        assert_eq!(Architecture::from_target(&t!("mipsle-unknown-linux-gnu"))?, Architecture::MipsLe);
+        Ok(())
+    }
+
+    #[test]
+    fn os_from_target() -> Result<()> {
+        assert_eq!(Os::from_target(&t!("x86_64-apple-darwin"))?, Os::Darwin);
+        assert_eq!(Os::from_target(&t!("x86_64-unknown-freebsd"))?, Os::Freebsd);
+        assert_eq!(Os::from_target(&t!("x86_64-unknown-netbsd"))?, Os::Netbsd);
+        assert_eq!(Os::from_target(&t!("sparcv9-sun-solaris"))?, Os::Solaris);
+        assert_eq!(Os::from_target(&t!("sparcv9-sun-illumos"))?, Os::Illumos);
+        assert_eq!(Os::from_target(&t!("aarch64-linux-android"))?, Os::Android);
+        assert_eq!(Os::from_target(&t!("x86_64-unknown-linux-gnu"))?, Os::Linux);
+        assert_eq!(Os::from_target(&t!("x86_64-pc-windows-msvc"))?, Os::Windows);
+        Ok(())
+    }
+}
```

Apply this and the code should work (there are 2 bug fixes and numerous tests).

Reply:
This code still needs to be formatted, however.